viewer/src/culling/ByVisibilityGpuSectorCuller.ts

```diff
@@ -0,0 +1,248 @@
+/*!
+ * Copyright 2020 Cognite AS
+ */
+
+// Note! We introduce ThreeJS as a dependency here, but it's not exposed.
+// If we add other rendering engines, we should consider to implement this in 'pure'
+// WebGL.
+import * as THREE from 'three';
+import { mat4 } from 'gl-matrix';
+
+import { WantedSector } from '../data/model/WantedSector';
+import { SectorScene } from '../models/cad/SectorScene';
+import { CadModel } from '../models/cad/CadModel';
+import { DetermineSectorsByProximityInput } from '../models/cad/determineSectors';
+import {
+  GpuOrderSectorsByVisibilityCoverage,
+  OrderSectorsByVisibilityCoverage
+} from '../views/threejs/OrderSectorsByVisibilityCoverage';
+import { SectorCuller } from './SectorCuller';
+import { TakenSectorTree } from './TakenSectorTree';
+import { PrioritizedWantedSector, DetermineSectorCostDelegate } from './types';
+import { fromThreeMatrix } from '../views/threejs/utilities';
+import { LevelOfDetail } from '../data/model/LevelOfDetail';
+import { SectorMetadata } from '../models/cad/types';
+
+/**
+ * Options for creating GpuBasedSectorCuller.
+ */
+export type ByVisibilityGpuSectorCullerOptions = {
+  /**
+   * Limit of how much data to load for a given view point. By default cost is measured in
+   * downloaded bytes per sector.
+   */
+  costLimit?: number;
+
+  /**
+   * Optional callback for determining the cost of a sector. The default unit of the cost
+   * function is bytes downloaded.
+   */
+  determineSectorCost?: DetermineSectorCostDelegate;
+
+  /**
+   * Sectors within this distance from the camera will always be loaded in high details.
+   */
+  highDetailProximityThreshold?: number;
+
+  /**
+   * Use a custom coverage utility to determine how "visible" each sector is.
+   */
+  coverageUtil?: OrderSectorsByVisibilityCoverage;
+};
+
+function assert(condition: boolean, message: string = 'assertion hit') {
+  // tslint:disable-next-line: no-console
+  console.assert(condition, message);
+}
+
+class TakenSectorMap {
+  get totalCost() {
```

Comment:
Why is this a getter, while getWantedSectorCount is a member function?

Reply:
It's maybe a stupid pattern, but I tend to use functions to indicate that there is some computational cost associated with the operation. Here getWantedSectorCount() is `O(M*N)` where M is models and N is sectors. `get totalCost` is `O(M)`.

I realize I'd have a hard time to make this argument in front of a jury and the natural follow-up-question would be "Why not clearly indicate this in the function name instead?" - and hence I will rename it to `determineWantedSectorCount()`